module/zfs/vdev_mirror.c

```diff
@@ -493,9 +513,34 @@ vdev_mirror_preferred_child_randomize(zio_t *zio)
 	return (mm->mm_preferred[p]);
 }
 
+static boolean_t
+vdev_mirror_child_readable(mirror_child_t *mc)
+{
+	vdev_t *vd = mc->mc_vd;
+
+	if (vd->vdev_top != NULL && vd->vdev_top->vdev_ops == &vdev_draid_ops)
+		return (vdev_draid_readable(vd, mc->mc_offset));
+	else
+		return (vdev_readable(vd));
+}
+
+static boolean_t
+vdev_mirror_child_missing(mirror_child_t *mc, uint64_t txg, uint64_t size)
+{
+	vdev_t *vd = mc->mc_vd;
+
+	if (vd->vdev_top != NULL && vd->vdev_top->vdev_ops == &vdev_draid_ops)
+		return (vdev_draid_missing(vd, mc->mc_offset, txg, size));
+	else
+		return (vdev_dtl_contains(vd, DTL_MISSING, txg, size));
+}
+
 /*
  * Try to find a vdev whose DTL doesn't contain the block we want to read
- * preferring vdevs based on determined load.
+ * preferring vdevs based on determined load. Distributed spares are an
+ * exception and are always preferred in order to detect gaps in the spare
+ * which can occur if another vdev fails.  They're also expected to
+ * outperform any single child which is good for performance.
```

Comment:
This was something which came up during our manual testing.  If you've done a proactive `zpool replace` to a distributed spare then both sides of the mirror will be online and have valid data.  Now if a different disk fails it will create gaps in the distributed spare.  This isn't a problem but in order to restore the redundancy we need to make sure those gaps get repaired when that failed disk is replaced.  In order to make sure that happens we need to read from the distributed spare during the resilver.  Otherwise, some fraction of the time we'll read good data from the fully intact half of the mirror and not detect we need to issue repair IO for the gap in the distributed spare.

This isn't the common case and it's pretty subtle.  I'll go ahead and add version of this explanation as a comment in the code.

Reply:
When you do the `zpool replace`, while it's in progress, only the physical disk has (all the) good data.  After the replace completes, don't we remove the replaced (physical) drive, leaving only the spare?  If that's true then there's no point where both children have (all the) valid data.

But assuming that my above analysis is wrong, wouldn't you have this problem in general, e.g. if you have a raidz and you've done `zpool replace` on one of its disks, and then you do a scrub/resilver of a different failed drive, if we don't read and check both sides of the `replace` mirror, we can miss and not repair some bad data.  But in the RAIDZ case, it would only be silent data corruption that we'd miss.  With DRAID I get that it could be due to a detected full-disk failure which "surprisingly" also takes out parts of the distributed spare, so the problem is more acute here.

Is this issue relevant to "healing resilver", "sequential resilver", or both?

Reply:
>  After the replace completes, don't we remove the replaced (physical) drive, leaving only the spare?

This is only the case is you're not replacing with a hot spare.  When replacing to a hot spare the original device  isn't automatically detached.  As far as I know this has been the intended behavior from day one.

> wouldn't you have this problem in general ... 

Yes, and in fact we do have this same issue with RAIDZ.  It's just a little more pronounced with a dRAID distributed spare since we can lose parts of the spare and not have it reflected in the DTLs.  With normal spare devices it's more all or nothing, you won't silently lose ranges of a disk.

> Is this issue relevant to "healing resilver", "sequential resilver", or both?

It applies to "healing resilver", "sequential resilver", and even scrub.  Any time there are replace/spare vdevs with multiple online children only one child will be read.  That data can only be checked at the RAIDZ/DRAID vdev since the replace/spare vdev level doesn't have a block pointer.  In practice, it's rarely an issue since this likely isn't a normal configuration but it is something I encountered when working through all the possible failure scenarios.  This is a long standing flaw in the current algorithm.  @mmaybee and I have discussed it and we have some ideas about how to address the general case.